File: java/src/main/java/com/varnaa/quickds/implementations/linkedlist/LinkedList.java
package com.varnaa.quickds.implementations.linkedlist;

import com.varnaa.quickds.implementations.prettyPrint.PrettyPrint;

import java.util.Random;

/**
 * @Author: swethavarnaa
 */
public class LinkedList implements PrettyPrint<Node> {
    private final int LENGTH = 10;
    private final Random RANDOM = new Random();
    private Node head;

    public LinkedList() {
        this.head = null;
    }

    public boolean isEmpty() {
        return this.head == null;
    }

    public void insertNode(int value) {
        if (isEmpty()) {
            head = new Node(value);
        } else {
            Node temp = head;
            while (temp.next != null) {
                temp = temp.next;
            }
            temp.next = new Node(value);
        }
    }

    public Node getHead() {
        return head;
    }

    public Node createRandomLinkedList() {
        LinkedList linkedList = new LinkedList();
        for (int i = 0; i < LENGTH; i++) {
            linkedList.insertNode(RANDOM.nextInt(10));
        }

        return linkedList.getHead();
    }


    @Override
    public void prettyPrint(Node head) {
        StringBuilder output = new StringBuilder();
        while (head != null) {
            output.append(head.value).append(" -> ");
        }
        output.append("null");
        System.out.println(output.toString());
    }
}

File: python/QuickDS/implementations/linkedlist.py
"""
@Author: rishi
"""
# Import required modules
import random

from .node import Node


# Build linked list class
class LinkedList:
    # Constructor
    def __init__(self):
        self.__head = None

    # Getter for head node
    def get_head(self):
        return self.__head

    # Check whether the linked list is empty
    def is_empty(self):
        return self.get_head() is None

    # Creates and inserts node into the linked list
    def insert_node(self, val):
        """
        :param val: value of the Node to be created
        """

        # if head Node is none then create the head Node with the value
        if self.is_empty():
            self.__head = Node(val)

        else:
            # create a copy of head Node
            temp = self.get_head()

            # traverse the available linked list
            while temp.get_next() is not None:
                temp = temp.get_next()

            # Add Node at the end of linked list
            temp.set_next(val)

    # Prints the linked list
    @staticmethod
    def pretty_print(headnode):
        """
        :param headnode: Head Node of linked list class instance
        """

        __str = ""
        head = headnode

        while head is not None:
            __str += str(head.get_val()) + "->"
            head = head.get_next()
        __str += "None"
        print(__str)

    # Creates a linked list with the values in array
    def create_linked_list(self, array):
        """
        :param array: iterable that specifies the values of linked list to be created
        :return: On success returns head node of linked list
        """

        # Traverse the array
        for i in array:
            self.insert_node(i)

        # Return success
        head = self.get_head()
        return head

    # Creates random linked list
    @staticmethod
    def create_random_linked_list(length=10):
        """
        :param length: length and range of the linked list
        :return: head node of the linked list

        """
        obj = LinkedList()
        for i in range(length):
            obj.insert_node(random.randint(1, length))
        return obj.get_head()
